### PR TITLE
Frontend refactorings & dashboard features

### PR DESCRIPTION
**Pull request by berekuk**
Ok, I got a bit carried away here (instead of working on graphql, going to start on that soon), but these are the cleanups I wanted to do eventually.

What's included:
1. Cleanups: coding style, extracted common components, enforced naming conventions, typescript. Also: initial `docs/coding-style.md` explaining what I consider to be best practice.
2. Simplified html/css.
3. Dashboards refactorings & new dashboard urls.

---

1) Cleanups:
- all component files are now named according to their primary component, e.g. MyComponent.tsx
- use named exports instead of default exports (it's easier not to mess rename refactorings this way, and also to type components properly)
- most components are typed with `React.FC<Props>`
- all components are rendered with JSX instead of function calls
- significant overhaul of `DisplayForecast` component, split footer in a separate file (and should've split more stuff), removed some unused code, etc.
- replaced most `let`s with `const`s wherever I've seen them :) (two more letters, but it's better for typing and clearer in intent)

2) Simplified HTML/CSS for `DisplayForecast`, `/tools` and `/dashboards`:
- I probably removed >70% of css classes in the code I changed; some of them weren't doing anything useful, and some were overly complicated
- Also extracted a few components, e.g. `Card`, `Button` and `LastUpdated`
- New code doesn't render _identically_, but all changes are minor (unless I messed up some rare combination of `showTimeStamp` and screen size... but I think I checked most of it, e.g. how question cards look on `/secretEmbed`)
- One aspect in which new code renders a bit differently is margins; it's generally a bad practice to mix margins with other code, so whenever I've encountered a few components with custom `mt-N mb-M` I tried to replace it with `space-y-N` in the parent component, and some sub-components shifted by a few pixels due to this

3) Dashboards:

- since I touched on /dashboards code, doing #17 seems relevant
- this lead me to url changes: dashboard pages are now at `/dashboards/view/[id]`, with redirects from the old `/dashboards?dashboardId=[id]` (using middleware, see `src/pages/_middleware.ts`)
- I could probably avoid the url change, but new schema seems better anyway (and the code was getting overly complicated with `initialVariable` and useEffects, I think trying to handle it the old way is too much of an anti-pattern, tbh)
- also fixed some bugs, e.g. "Create dashboard" button rendering was broken during dashboard's creation, should be better now

### CODE REVIEW

File: src/web/display/DisplayForecast/ForecastFooter.tsx [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+  platformLabel,
+  numforecasts,
+  qualityindicators,
+  timestamp,
+  showTimeStamp,
+  expandFooterToFullWidth,
+}) => {
+  // I experimented with justify-evenly, justify-around, etc., here: https://tailwindcss.com/docs/justify-content
+  // I came to the conclusion that as long as the description isn't justified too, aligning the footer symmetrically doesn't make sense
+  // because the contrast is jarring.
```
Comment (NunoSempere) [inline src/web/display/DisplayForecast/ForecastFooter.tsx:258]: Not sure how I feel about this; for me, non-symmetric footers feel really jarring. But I'll defer to you.

File: docs/coding-style.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+## Typescript and React/Next
+
+- use `React.FC<Props>` type for React components, e.g. `const MyComponent: React.FC<Props> = ({ ... }) => { ... };`
+- use `NextPage<Props>` for typing stuff in `src/pages/`
+- use generic versions of `GetServerSideProps<Props>` and `GetStaticProps<Props>`
+
+# React
+
+- create one file per one component (tiny helper components in the same file are fine)
+- if
```
Comment (NunoSempere) [inline docs/coding-style.md]: if?

File: docs/coding-style.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+# Styles
+
+- use [Tailwind](https://tailwindcss.com/)
+- avoid positioning styles in components, position elements from the outside (e.g. with [space-\*](https://tailwindcss.com/docs/space) or grid/flexbox)
+
+# General notes
+
+- use `const` instead of `let` whenever possible
+- set up [prettier](https://prettier.io/) to format code on save
```
Comment (NunoSempere) [inline docs/coding-style.md:29]: Thanks for the documenting!

**Comment by NunoSempere** [comment]
This looks good. Could you also create a dashboardEmbed with the same schema (dashboardEmbed/view/[id]), to replace secretDashboard, but without deleting the secretDashboard page yet?

File: src/web/display/DisplayForecast/ForecastFooter.tsx (hunk 1 above)
Comment (berekuk) [reply src/web/display/DisplayForecast/ForecastFooter.tsx:258]: Uh, that's your comment that I should've removed :)
I haven't changed anything re: footers alignment.

File: docs/coding-style.md (hunk 2 above)
Comment (berekuk) [reply docs/coding-style.md]: Already fixed in 8d6cf48b4116194ead2886fd13571c518e92f2a4

**Comment by berekuk** [comment]
> Could you also create a dashboardEmbed with the same schema (dashboardEmbed/view/[id]), to replace secretDashboard, but without deleting the secretDashboard page yet?

Maybe `dashboard/embed/[id]` for namespacing then? Or `/dashboard/view/[id]/embed`.

I could just add a redirect to it and keep `/secretDashboard` forever as legacy.

Btw, where is `/secretDashboard` used currently? I see by Netlify analytics that it gets a lot of traffic (from globalguessing.com, probably?), but I couldn't find any pages where it's embedded. Reason I'm asking: it dumps a raw json into html, and I wonder if all pages that embed it are currently messed up.

**Comment by NunoSempere** [comment]
> Maybe dashboard/embed/[id] for namespacing then? Or /dashboard/view/[id]/embed.

dashboard/embed/[id] sounds a bit better.

> I could just add a redirect to it and keep /secretDashboard forever as legacy.

Sounds good, though maybe not forever.

> Btw, where is /secretDashboard used currently?

Here: https://globalguessing.com/russia-ukraine-forecasts/, that I know of

**Comment by berekuk** [comment]
> Reason I'm asking: it dumps a raw json into html, and I wonder if all pages that embed it are currently messed up.

Oh, I mixed it up with /secretEmbed. /secretDashboard is fine.

File: src/web/display/DisplayForecast/ForecastFooter.tsx (hunk 1 above)
Comment (NunoSempere) [reply src/web/display/DisplayForecast/ForecastFooter.tsx:258]: Oof, lol

**Comment by NunoSempere** [comment]
/secretEmbed is used by the twitter bot.

Is this PR ready to be merged?

**Comment by berekuk** [comment]
> Is this PR ready to be merged?

Yes! ✔️ 

Also, I added `/dashboards/embed/[id]` redirect in the final commit.

**Comment by NunoSempere** [comment]
I'm getting some errors in Vercel related to this commit @berekuk 

temp

**Comment by NunoSempere** [comment]
> Also, I added /dashboards/embed/[id] redirect in the final commit.

Nice

**Comment by berekuk** [comment]
> I'm getting some errors in Vercel related to this commit

That's #15 again (due to me pushing longer versions of infer/gjopen cookies there). Fixed.

**Pull request merged**